src/jenkins.py

```diff
@@ -705,3 +709,195 @@ def get_agent_name(unit_name: str) -> str:
         The agent node name registered on Jenkins server.
     """
     return unit_name.replace("/", "-")
+
+
+PLUGIN_NAME_GROUP = r"^([a-zA-Z0-9-_]+)"
+WHITESPACE = r"\s*"
+VERSION_GROUP = r"\((.*?)\)"
+DEPENDENCIES_GROUP = r"\[(.*?)\]"
+PLUGIN_CAPTURE = rf"{PLUGIN_NAME_GROUP}{WHITESPACE}{VERSION_GROUP}"
+PLUGIN_LINE_CAPTURE = rf"{PLUGIN_CAPTURE} => {DEPENDENCIES_GROUP}"
+
+
+def _get_plugin_name(plugin_info: str) -> str:
+    """Get plugin name given a plugin info string of format <plugin-shortname> (<plugin-version>).
+
+    Args:
+        plugin_info: Text containing plugin name and plugin version.
+
+    Raises:
+        ValidationError: if the plugin info string does not conform to expected format.
+
+    Returns:
+        The plugin shortname.
+    """
+    match = re.match(PLUGIN_CAPTURE, plugin_info)
+    if not match:
+        raise ValidationError(f"No plugin matched in: {plugin_info}")
+    return match.group(1)
+
+
+def _build_dependencies_lookup(
+    plugin_dependency_outputs: typing.Iterable[str],
+) -> dict[str, tuple[str, ...]]:
+    """Build a lookup table of plugin short name to list of dependency plugin's short names.
+
+    Args:
+        plugin_dependency_outputs: The plugin dependency output from Jenkins Groovy script.
+
+    Returns:
+        The dependency lookup table.
+    """
+    # coverage tool thinks defaultdict() has an exit condition.
+    dependency_lookup: dict[str, tuple[str, ...]] = defaultdict(lambda: ())  # pragma: no cover
```

Comment:
I think you can do `defaultdict(tuple)`

Reply:
And, if we're not using the default diction functionality, why use it please?

Reply:
I think it is *implicitly* getting used when the code traverses the `dependency_lookup` recursively. It can help be defensive about the returned value from Jenkins, although I trust the jenkins output.
For example: 
L803: `for dependency in dependency_lookup[plugin]:` might throw a `KeyError` on a regular dict, but with `defaultdict` it helps just skip the traversal of plugins that are not found.

Reply:
Refactored to use normal dict and handle KeyError with warnings